src/util/file_io.c

```diff
@@ -0,0 +1,67 @@
+/******************************************************************************
+ *
+ * Copyright 2018 Xaptum, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+ #include <stdint.h>
+ #include <stdio.h>
+ #include <xtt/util/file_io.h>
+
+ int save_to_file(unsigned char *buffer, size_t buffer_length, const char *filename)
+ {
+     FILE *file_ptr = fopen(filename, "wb");
+
+     if (NULL == file_ptr){
+         fprintf(stderr, "Could not find file: '%s'\n", filename);
+         return -1;
+     }
+
+     size_t bytes_written = fwrite(buffer, 1, buffer_length, file_ptr);
+
+     if (buffer_length != bytes_written) {
+         if(0 != ferror(file_ptr)){
+             fprintf(stderr, "Error writing to file '%s'\n", filename);
+             return -1;
+         } else if (0 != feof(file_ptr)){
+             return (int) bytes_written;
+         }
+     }
+
+     (void)fclose(file_ptr);
+     return (int) bytes_written;
+ }
+
+ int read_from_file(const char *filename, unsigned char *buffer, size_t bytes_to_read)
+ {
+     FILE *file_ptr = fopen(filename, "rb");
+
+     if (NULL == file_ptr){
+         fprintf(stderr, "Could not find file: '%s'\n", filename);
+         return -1;
+     }
+     size_t bytes_read = fread(buffer, 1, bytes_to_read, file_ptr);
+     if (bytes_to_read != bytes_read) {
+         if(0 != ferror(file_ptr)){
+             fprintf(stderr, "Error writing to file '%s'\n", filename);
+             return -1;
+         } else if (0 != feof(file_ptr)){
+             return (int) bytes_read;
+         }
+     }
+
+     (void)fclose(file_ptr);
```

Comment:
The `goto`s for `file_io` errors look good, but we should just `return` if `fopen` fails, rather than going to `cleanup`. That's because we don't want to call `fclose` on a `NULL` `FILE` pointer.